adam/learner/cross_situational.py

```diff
@@ -0,0 +1,567 @@
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescription]
+    ) -> None:
+        logging.info("Observation %s", self._observation_num)
+        self._observation_num += 1
+
+        perception = learning_example.perception
+        if len(perception.frames) != 1:
+            raise RuntimeError(
+                "Cross-situational learner can only handle single frames for now"
+            )
+        if isinstance(perception.frames[0], DevelopmentalPrimitivePerceptionFrame):
+            original_perception_graph = PerceptionGraph.from_frame(
+                perception.frames[0]
+            ).copy_as_digraph()
+        else:
+            raise RuntimeError("Cannot process perception type.")
+        # Remove learner from the perception
+        observed_perception_graph = graph_without_learner(original_perception_graph)
+        observed_linguistic_description = (
+            learning_example.linguistic_description.as_token_sequence()
+        )
+
+        self.learn_with_cross_situational(
+            observed_perception_graph, observed_linguistic_description
+        )
+
+    # Re-implement for cross-situational
+    def learn_with_cross_situational(
+        self,
+        observed_perception_graph: PerceptionGraph,
+        observed_linguistic_description: Tuple[str, ...],
+    ) -> None:
+        logging.info(
+            f"Cross-situational learner observing {observed_linguistic_description}"
+        )
+        # The learner’s words are W, meanings are M, their associations are A, and the new
+        # utterance is U = (W_U, M_U).
+        # For every w in W_U
+        # Adding "dummy word" that smoothes out learning for situations where a word shows up
+        # without it's meaning or vice versa. Also smoothes out meanings that show up without a word
+        # in general.
+        words = observed_linguistic_description + tuple("d")
+        meanings = [
+            PerceptionGraphPattern.from_graph(object_).perception_graph_pattern
+            for object_ in self.get_objects_from_perception(observed_perception_graph)
+        ]
+        for word in words:
+            # TODO: pursuit learner hard-coded to ignore determiners
+            # https://github.com/isi-vista/adam/issues/498
+            if word in ("a", "the"):
+                continue
+            if word in self._words_to_number_of_observations:
+                self._words_to_number_of_observations[word] += 1
+            else:
+                self._words_to_number_of_observations[word] = 1
+
+            # If word is not known, go through learning steps
+            if word not in self._lexicon:
+                logging.info(f"Considering '{word}'")
+                # Initialize necessary components if needed
+                if word not in self._words_to_hypotheses_and_scores:
+                    self._words_to_hypotheses_and_scores[word] = {}
+                    self._words_to_hypotheses_and_probability[word] = {}
```

Comment:
You can use `defaultdict` class to avoid having to initailize empty dicts: https://docs.python.org/3/library/collections.html#collections.defaultdict